lightning/src/offers/invoice_error.rs

```diff
@@ -0,0 +1,219 @@
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your option.
+// You may not use this file except in accordance with one or both of these
+// licenses.
+
+//! Data structures and encoding for `invoice_error` messages.
+
+use core::convert::TryFrom;
+use crate::io;
+use crate::offers::parse::{ParseError, ParsedMessage, SemanticError};
+use crate::util::ser::{HighZeroBytesDroppedBigSize, WithoutLength, Writeable, Writer};
+
+use crate::prelude::*;
+
+/// An error in response to an [`InvoiceRequest`] or an [`Invoice`].
+///
+/// [`InvoiceRequest`]: crate::offers::invoice_request::InvoiceRequest
+/// [`Invoice`]: crate::offers::invoice::Invoice
+#[derive(Clone, Debug)]
+#[cfg_attr(test, derive(PartialEq))]
+pub struct InvoiceError {
+	erroneous_field: Option<ErroneousField>,
+	message: String,
+}
+
+/// The field in the [`InvoiceRequest`] or the [`Invoice`] that contained an error.
+///
+/// [`InvoiceRequest`]: crate::offers::invoice_request::InvoiceRequest
+/// [`Invoice`]: crate::offers::invoice::Invoice
+#[derive(Clone, Debug)]
+#[cfg_attr(test, derive(PartialEq))]
+pub struct ErroneousField {
+	tlv_fieldnum: u64,
+	suggested_value: Option<Vec<u8>>,
+}
+
+impl InvoiceError {
+	/// The field in the [`InvoiceRequest`] or the [`Invoice`] that contained an error.
+	///
+	/// [`InvoiceRequest`]: crate::offers::invoice_request::InvoiceRequest
+	/// [`Invoice`]: crate::offers::invoice::Invoice
+	pub fn erroneous_field(&self) -> Option<&ErroneousField> {
+		self.erroneous_field.as_ref()
+	}
+
+	/// An explanation of the error.
+	pub fn message(&self) -> &str {
+		&self.message
+	}
+
+	pub(super) fn as_tlv_stream(&self) -> InvoiceErrorTlvStreamRef {
+		InvoiceErrorTlvStreamRef {
+			erroneous_field: self.erroneous_field.as_ref().map(|f| f.tlv_fieldnum),
+			suggested_value: self.erroneous_field.as_ref().and_then(|f| f.suggested_value.as_ref()),
+			error: Some(&self.message),
+		}
+	}
+}
+
+impl core::fmt::Display for InvoiceError {
+	fn fmt(&self, f: &mut core::fmt::Formatter) -> Result<(), core::fmt::Error> {
+		f.write_str(self.message())
+	}
+}
+
+impl Writeable for InvoiceError {
+	fn write<W: Writer>(&self, writer: &mut W) -> Result<(), io::Error> {
+		self.as_tlv_stream().write(writer)
+	}
+}
+
+tlv_stream!(InvoiceErrorTlvStream, InvoiceErrorTlvStreamRef, 0..=5, {
+	(1, erroneous_field: (u64, HighZeroBytesDroppedBigSize)),
+	(3, suggested_value: (Vec<u8>, WithoutLength)),
+	(5, error: (String, WithoutLength)),
+});
+
+impl TryFrom<Vec<u8>> for InvoiceError {
+	type Error = ParseError;
+
+	fn try_from(bytes: Vec<u8>) -> Result<Self, Self::Error> {
+		let invoice_error = ParsedMessage::<InvoiceErrorTlvStream>::try_from(bytes)?;
+		let ParsedMessage { tlv_stream, .. } = invoice_error;
+		let InvoiceErrorTlvStream { erroneous_field, suggested_value, error } = tlv_stream;
+
+		let erroneous_field = match (erroneous_field, suggested_value) {
+			(None, None) => None,
+			(None, Some(_)) => None,
```

Comment:
Yeah, though the spec is underspecified. I left a comment: https://github.com/lightning/bolts/pull/798/files#r1200840518

I wonder if we should just ignore `erroneous_field` and `suggested_value` since they are odd and not very useful at the moment.